subprojects/snapshots/src/main/java/org/gradle/internal/vfs/impl/WatchingVirtualFileSystem.java

```diff
@@ -147,65 +162,73 @@ public void handleLostState() {
                     stopWatching();
                 }
             });
-            getRoot().update(SnapshotHierarchy::empty);
-            delegatingUpdateFunctionDecorator.setSnapshotDiffListener(snapshotDiffListener);
+            delegatingUpdateFunctionDecorator.setSnapshotDiffListener(snapshotDiffListener, this::handleWatcherChangeErrors);
             long endTime = System.currentTimeMillis() - startTime;
             LOGGER.warn("Spent {} ms registering watches for file system events", endTime);
+            // TODO: Move start watching early enough so that the root is always empty
+            return currentRoot.empty();
         } catch (Exception ex) {
             LOGGER.error("Couldn't create watch service, not tracking changes between builds", ex);
-            invalidateAll();
-            close();
+            closeUnderLock();
         }
+        return currentRoot.empty();
     }
 
-    private void handleWatcherChanges(Consumer<FileWatcherRegistry> consumer) {
+    private SnapshotHierarchy handleWatcherChangeErrors(SnapshotHierarchy currentRoot, Runnable runnable) {
         try {
-            consumer.accept(watchRegistry);
+            runnable.run();
+            return currentRoot;
         } catch (WatchingNotSupportedException ex) {
             // No stacktrace here, since this is a known shortcoming of our implementation
             LOGGER.warn("Watching not supported, not tracking changes between builds: {}", ex.getMessage());
-            stopWatching();
+            return stopWatching(currentRoot);
         } catch (Exception ex) {
             LOGGER.error("Couldn't update watches, not watching anymore", ex);
-            stopWatching();
+            return stopWatching(currentRoot);
         }
     }
 
     /**
      * Stop watching the known areas of the file system, and invalidate
-     * the parts that have been changed since calling {@link #startWatching()}}.
+     * the parts that have been changed since calling {@link #startWatching(SnapshotHierarchy)}}.
      */
     private void stopWatching() {
-        updateWatchRegistry(fileWatcherRegistry -> {
+        getRoot().update(this::stopWatching);
+    }
```

Comment:
This confused me a bit, wonder if we could call `stopWatching()` here something else?